Gradle listExtension - fails on new release or project out of date with latest release

### Description

I got this once  2.1.2 was released.

* What went wrong:
Execution failed for task ':listExtensions'.
> Unable to list extensions

Other tasks work with older version.

Work-around:
Manually update `gradle.properties` to reflect latest version.


### Implementation ideas

_No response_

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':listExtensions'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$3(ExecuteActionsTaskExecuter.java:186)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:268)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:184)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:173)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:411)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:398)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:391)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:377)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: org.gradle.api.GradleException: Unable to list extensions
        at io.quarkus.gradle.tasks.QuarkusListExtensions.listExtensions(QuarkusListExtensions.java:122)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:58)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:51)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:29)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$2.run(ExecuteActionsTaskExecuter.java:494)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:479)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:462)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$400(ExecuteActionsTaskExecuter.java:105)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.executeWithPreviousOutputFiles(ExecuteActionsTaskExecuter.java:273)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:251)
        at org.gradle.internal.execution.steps.ExecuteStep.lambda$executeOperation$1(ExecuteStep.java:66)
        at org.gradle.internal.execution.steps.ExecuteStep.executeOperation(ExecuteStep.java:66)
        at org.gradle.internal.execution.steps.ExecuteStep.access$000(ExecuteStep.java:34)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:47)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:44)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:44)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:34)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:72)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:42)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:53)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:39)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:44)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:77)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:58)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:54)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:32)
        at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:57)
        at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:38)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:63)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:30)
        at org.gradle.internal.execution.steps.BuildCacheStep.executeWithoutCache(BuildCacheStep.java:176)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:76)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:47)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:43)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:32)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:39)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:25)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:102)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:95)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:55)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:39)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:83)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:44)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:27)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:96)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:52)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:83)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:54)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:74)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:88)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:88)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:34)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:38)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:46)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:34)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.lambda$execute$0(AssignWorkspaceStep.java:43)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution$3.withWorkspace(ExecuteActionsTaskExecuter.java:286)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:43)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:33)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:40)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:30)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:54)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:40)
        at org.gradle.internal.execution.impl.DefaultExecutionEngine.rebuild(DefaultExecutionEngine.java:46)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$0(ExecuteActionsTaskExecuter.java:182)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:182)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:173)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:411)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:398)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:391)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:377)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.RuntimeException: Failed to resolve extensions catalog
        at io.quarkus.gradle.tasks.QuarkusPlatformTask.extensionsCatalog(QuarkusPlatformTask.java:54)
        at io.quarkus.gradle.tasks.QuarkusPlatformTask.getQuarkusProject(QuarkusPlatformTask.java:134)
        at io.quarkus.gradle.tasks.QuarkusListExtensions.listExtensions(QuarkusListExtensions.java:95)
        ... 123 more
Caused by: java.lang.IllegalArgumentException: No catalogs provided
        at io.quarkus.registry.catalog.json.JsonCatalogMerger.merge(JsonCatalogMerger.java:21)
        at io.quarkus.registry.ExtensionCatalogResolver$ExtensionCatalogBuilder.build(ExtensionCatalogResolver.java:334)
        at io.quarkus.registry.ExtensionCatalogResolver.resolveExtensionCatalog(ExtensionCatalogResolver.java:414)
        at io.quarkus.registry.ExtensionCatalogResolver.resolveExtensionCatalog(ExtensionCatalogResolver.java:597)
        at io.quarkus.gradle.tasks.QuarkusPlatformTask.extensionsCatalog(QuarkusPlatformTask.java:52)
        ... 125 more

--- a/independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/registry/client/TestRegistryClientBuilder.java
+++ b/independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/registry/client/TestRegistryClientBuilder.java
@@ -572,7 +572,7 @@ public class TestRegistryClientBuilder {
         }
     }
 
-    static Path getRegistryMemberCatalogPath(Path registryDir, ArtifactCoords bom) {
+    public static Path getRegistryMemberCatalogPath(Path registryDir, ArtifactCoords bom) {
         return getMemberCatalogPath(registryDir.resolve("members"), bom);
     }
 
@@ -605,7 +605,7 @@ public class TestRegistryClientBuilder {
         return registryDir.resolve("config.json");
     }
 
-    static Path getRegistryDir(Path baseDir, String registryId) {
+    public static Path getRegistryDir(Path baseDir, String registryId) {
         return baseDir.resolve(registryId);
     }
 

--- a/independent-projects/tools/devtools-testing/src/test/java/io/quarkus/devtools/project/create/ExtensionCatalogFromProvidedPlatformsTest.java
+++ b/independent-projects/tools/devtools-testing/src/test/java/io/quarkus/devtools/project/create/ExtensionCatalogFromProvidedPlatformsTest.java
@@ -1,0 +1,114 @@
+package io.quarkus.devtools.project.create;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.quarkus.bootstrap.resolver.maven.workspace.ModelUtils;
+import io.quarkus.devtools.testing.registry.client.TestRegistryClientBuilder;
+import io.quarkus.maven.ArtifactCoords;
+import io.quarkus.registry.ExtensionCatalogResolver;
+import io.quarkus.registry.catalog.ExtensionCatalog;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.stream.Collectors;
+import org.apache.maven.model.Dependency;
+import org.apache.maven.model.Model;
+import org.assertj.core.util.Arrays;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+public class ExtensionCatalogFromProvidedPlatformsTest extends MultiplePlatformBomsTestBase {
+
+    private static final String MAIN_PLATFORM_KEY = "org.acme.platform";
+
+    @BeforeAll
+    public static void setup() throws Exception {
+        TestRegistryClientBuilder.newInstance()
+                //.debug()
+                .baseDir(configDir())
+                // registry
+                .newRegistry("registry.acme.org")
+                // platform key
+                .newPlatform(MAIN_PLATFORM_KEY)
+                // 2.0 STREAM
+                .newStream("2.0")
+                // 2.0.4 release
+                .newRelease("2.0.4")
+                .quarkusVersion("2.2.2")
+                // default bom including quarkus-core + essential metadata
+                .addCoreMember()
+                // foo platform member
+                .newMember("acme-foo-bom").addExtension("acme-foo").release()
+                .stream().platform()
+                // 1.0 STREAM
+                .newStream("1.0")
+                // 1.0.1 release
+                .newRelease("1.0.1")
+                .quarkusVersion("1.1.2")
+                .addCoreMember()
+                .newMember("acme-foo-bom").addExtension("acme-foo").release()
+                .newMember("acme-baz-bom").addExtension("acme-baz").release()
+                .registry()
+                .newNonPlatformCatalog("1.1.2")
+                .addExtension("org.acme", "acme-quarkus-other", "5.5.5")
+                .registry()
+                .clientBuilder()
+                .build();
+
+        enableRegistryClient();
+
+        // install 1.0.0 version which is not recommended by the registry any more
+        installNotRecommendedVersion("1.0.1", "1.0.0", "acme-foo-bom");
+        installNotRecommendedVersion("1.0.1", "1.0.0", "acme-baz-bom");
+        installNotRecommendedVersion("1.0.1", "1.0.0", "quarkus-bom");
+    }
+
+    private static void installNotRecommendedVersion(String baseRecommendedVersion, String nonRecommendedVersion,
+            String memberArtifactId) throws IOException {
+        Path repoDir = TestRegistryClientBuilder.getMavenRepoDir(configDir());
+        Path p = repoDir;
+        for (String s : MAIN_PLATFORM_KEY.split("\\.")) {
+            p = p.resolve(s);
+        }
+        final Path groupIdDir = p;
+
+        p = groupIdDir.resolve(memberArtifactId).resolve(baseRecommendedVersion)
+                .resolve(memberArtifactId + "-" + baseRecommendedVersion + ".pom");
+        final Model model = ModelUtils.readModel(p);
+        model.setVersion(nonRecommendedVersion);
+        for (Dependency d : model.getDependencyManagement().getDependencies()) {
+            if (baseRecommendedVersion.equals(d.getVersion())) {
+                d.setVersion(nonRecommendedVersion);
+            }
+        }
+        p = p.getParent().getParent().resolve(nonRecommendedVersion)
+                .resolve(memberArtifactId + "-" + nonRecommendedVersion + ".pom");
+        Files.createDirectories(p.getParent());
+        ModelUtils.persistModel(p, model);
+
+        p = TestRegistryClientBuilder.getRegistryMemberCatalogPath(
+                TestRegistryClientBuilder.getRegistryDir(configDir(), "registry.acme.org"),
+                new ArtifactCoords(MAIN_PLATFORM_KEY, memberArtifactId, null, "pom", baseRecommendedVersion));
+        final String jsonContent = Files.readString(p).replace(baseRecommendedVersion, nonRecommendedVersion);
+        String jsonName = p.getFileName().toString().replace(baseRecommendedVersion, nonRecommendedVersion);
+        Files.writeString(p.getParent().resolve(jsonName), jsonContent);
+    }
+
+    protected String getMainPlatformKey() {
+        return MAIN_PLATFORM_KEY;
+    }
+
+    @Test
+    public void test() throws Exception {
+        final ExtensionCatalog catalog = ExtensionCatalogResolver.builder().build().resolveExtensionCatalog(
+                Collections.singletonList(ArtifactCoords.fromString(MAIN_PLATFORM_KEY + ":acme-baz-bom::pom:1.0.0")));
+        assertThat(Arrays.asList(new ArtifactCoords[] {
+                ArtifactCoords.fromString("io.quarkus:quarkus-core:1.1.2"),
+                ArtifactCoords.fromString(MAIN_PLATFORM_KEY + ":acme-foo:1.0.0"),
+                ArtifactCoords.fromString(MAIN_PLATFORM_KEY + ":acme-baz:1.0.0"),
+                ArtifactCoords.fromString("org.acme:acme-quarkus-other:5.5.5")
+        })).isEqualTo(
+                catalog.getExtensions().stream().map(e -> e.getArtifact()).collect(Collectors.toList()));
+    }
+}

--- a/independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/ExtensionCatalogResolver.java
+++ b/independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/ExtensionCatalogResolver.java
@@ -310,6 +310,7 @@ public class ExtensionCatalogResolver {
         private final List<ExtensionCatalog> catalogs = new ArrayList<>();
         final Map<String, List<RegistryExtensionResolver>> registriesByQuarkusCore = new HashMap<>();
         private final Map<String, Integer> compatibilityCodes = new LinkedHashMap<>();
+        final List<String> upstreamQuarkusVersions = new ArrayList<>(1);
 
         void addCatalog(ExtensionCatalog c) {
             catalogs.add(c);
@@ -399,10 +400,9 @@ public class ExtensionCatalogResolver {
     private void addOriginPreference(final ExtensionCatalog ec, OriginPreference originPreference) {
         Map<String, Object> metadata = ec.getMetadata();
         if (metadata.isEmpty()) {
-            metadata = new HashMap<>();
+            metadata = new HashMap<>(1);
             ((JsonExtensionCatalog) ec).setMetadata(metadata);
         }
-
         metadata.put("origin-preference", originPreference);
     }
 
@@ -416,17 +416,16 @@ public class ExtensionCatalogResolver {
             throw new RegistryResolutionException("No registries configured");
         }
 
-        return resolveExtensionCatalog(quarkusCoreVersion, new ExtensionCatalogBuilder(), Collections.emptyMap());
+        return resolveExtensionCatalog(quarkusCoreVersion, new ExtensionCatalogBuilder(), Collections.emptySet());
     }
 
     private ExtensionCatalog resolveExtensionCatalog(String quarkusCoreVersion,
-            final ExtensionCatalogBuilder catalogBuilder, Map<String, Platform> preferredPlatforms)
+            final ExtensionCatalogBuilder catalogBuilder, Set<String> preferredPlatforms)
             throws RegistryResolutionException {
-        final List<String> upstreamQuarkusVersions = new ArrayList<>(1);
-        collectPlatforms(quarkusCoreVersion, catalogBuilder, upstreamQuarkusVersions, preferredPlatforms);
+        collectPlatforms(quarkusCoreVersion, catalogBuilder, preferredPlatforms);
         int i = 0;
-        while (i < upstreamQuarkusVersions.size()) {
-            collectPlatforms(upstreamQuarkusVersions.get(i++), catalogBuilder, upstreamQuarkusVersions, preferredPlatforms);
+        while (i < catalogBuilder.upstreamQuarkusVersions.size()) {
+            collectPlatforms(catalogBuilder.upstreamQuarkusVersions.get(i++), catalogBuilder, preferredPlatforms);
         }
         return catalogBuilder.build();
     }
@@ -528,17 +527,17 @@ public class ExtensionCatalogResolver {
     }
 
     @SuppressWarnings("unchecked")
-    public ExtensionCatalog resolveExtensionCatalog(Collection<ArtifactCoords> platforms)
+    public ExtensionCatalog resolveExtensionCatalog(Collection<ArtifactCoords> preferredPlatforms)
             throws RegistryResolutionException {
-        if (platforms.isEmpty()) {
+        if (preferredPlatforms.isEmpty()) {
             return resolveExtensionCatalog();
         }
 
         final ExtensionCatalogBuilder catalogBuilder = new ExtensionCatalogBuilder();
-        final Map<String, Platform> preferredPlatforms = new LinkedHashMap<>();
+        final Set<String> preferredPlatformKeys = new HashSet<>();
         String quarkusVersion = null;
         int platformIndex = 0;
-        for (ArtifactCoords bom : platforms) {
+        for (ArtifactCoords bom : preferredPlatforms) {
             final List<RegistryExtensionResolver> registries;
             try {
                 registries = filterRegistries(r -> r.checkPlatform(bom));
@@ -554,8 +553,30 @@ public class ExtensionCatalogResolver {
                 throw new RuntimeException(buf.toString());
             }
 
-            final ExtensionCatalog catalog = resolvePlatformExtensions(bom, registries);
+            if (registries.isEmpty()) {
+                log.warn("None of the configured registries recognizes platform %s", bom);
+                continue;
+            }
+
+            ExtensionCatalog catalog = null;
+            RegistryExtensionResolver registry = null;
+            for (int i = 0; i < registries.size(); ++i) {
+                registry = registries.get(i);
+                try {
+                    catalog = registry.resolvePlatformExtensions(bom);
+                    break;
+                } catch (RegistryResolutionException e) {
+                }
+            }
+
             if (catalog == null) {
+                final StringBuilder buf = new StringBuilder();
+                buf.append("Failed to resolve platform ").append(bom).append(" using the following registries: ");
+                buf.append(registries.get(0).getId());
+                for (int i = 1; i < registries.size(); ++i) {
+                    buf.append(", ").append(registries.get(i++));
+                }
+                log.warn(buf.toString());
                 continue;
             }
 
@@ -571,12 +592,11 @@ public class ExtensionCatalogResolver {
                     o = md.get("platform-key");
                     if (o != null && o instanceof String) {
                         final String platformKey = o.toString();
-                        if (preferredPlatforms.containsKey(platformKey)) {
+                        if (!preferredPlatformKeys.add(platformKey)) {
                             continue;
                         }
                         final JsonPlatform p = new JsonPlatform();
                         p.setPlatformKey(platformKey);
-                        preferredPlatforms.put(platformKey, p);
 
                         final JsonPlatformStream stream = new JsonPlatformStream();
                         stream.setId(String.valueOf(md.getOrDefault("stream", "default")));
@@ -598,20 +618,21 @@ public class ExtensionCatalogResolver {
                             }
                             release.setMemberBoms(coords);
                         }
+
+                        collectPlatforms(quarkusVersion, catalogBuilder, registry, platformIndex,
+                                p);
                         continue;
                     }
                 }
             }
-
-            catalogBuilder.addCatalog(catalog);
-
             final OriginPreference originPreference = new OriginPreference(0, ++platformIndex, 1, 1,
                     catalogBuilder.getCompatibilityCode(quarkusVersion));
             addOriginPreference(catalog, originPreference);
+            catalogBuilder.addCatalog(catalog);
         }
 
         return preferredPlatforms.isEmpty() ? catalogBuilder.build()
-                : resolveExtensionCatalog(quarkusVersion, catalogBuilder, preferredPlatforms);
+                : resolveExtensionCatalog(quarkusVersion, catalogBuilder, preferredPlatformKeys);
     }
 
     public void clearRegistryCache() throws RegistryResolutionException {
@@ -625,27 +646,6 @@ public class ExtensionCatalogResolver {
         if (registriesTotal == 0) {
             throw new RegistryResolutionException("No registries configured");
         }
-    }
-
-    private ExtensionCatalog resolvePlatformExtensions(ArtifactCoords bom, List<RegistryExtensionResolver> registries) {
-        if (registries.isEmpty()) {
-            log.debug("None of the configured registries recognizes platform %s", bom);
-            return null;
-        }
-        for (RegistryExtensionResolver registry : registries) {
-            try {
-                return registry.resolvePlatformExtensions(bom);
-            } catch (RegistryResolutionException e) {
-            }
-        }
-        final StringBuilder buf = new StringBuilder();
-        buf.append("Failed to resolve platform ").append(bom).append(" using the following registries: ");
-        buf.append(registries.get(0).getId());
-        for (int i = 1; i < registries.size(); ++i) {
-            buf.append(", ").append(registries.get(i++));
-        }
-        log.warn(buf.toString());
-        return null;
     }
 
     private void appendNonPlatformExtensions(
@@ -684,9 +684,8 @@ public class ExtensionCatalogResolver {
         throw new IllegalStateException(buf.toString());
     }
 
-    private void collectPlatforms(String quarkusCoreVersion,
-            ExtensionCatalogBuilder catalogBuilder,
-            Collection<String> upstreamQuarkusVersions, Map<String, Platform> preferredPlatforms)
+    private void collectPlatforms(String quarkusCoreVersion, ExtensionCatalogBuilder catalogBuilder,
+            Set<String> preferredPlatforms)
             throws RegistryResolutionException {
         final List<RegistryExtensionResolver> quarkusVersionRegistries = catalogBuilder
                 .getRegistriesForQuarkusCore(quarkusCoreVersion);
@@ -700,25 +699,19 @@ public class ExtensionCatalogResolver {
             if (platforms.isEmpty()) {
                 continue;
             }
-            int platformIndex = 0;
-            for (Platform p : preferredPlatforms.values()) {
-                ++platformIndex;
-                collectPlatforms(quarkusCoreVersion, catalogBuilder, upstreamQuarkusVersions, registry, platformIndex,
-                        p);
-            }
+            int platformIndex = preferredPlatforms.size();
             for (Platform p : platforms) {
-                if (preferredPlatforms.containsKey(p.getPlatformKey())) {
+                if (preferredPlatforms.contains(p.getPlatformKey())) {
                     continue;
                 }
                 ++platformIndex;
-                collectPlatforms(quarkusCoreVersion, catalogBuilder, upstreamQuarkusVersions, registry, platformIndex,
-                        p);
+                collectPlatforms(quarkusCoreVersion, catalogBuilder, registry, platformIndex, p);
             }
         }
     }
 
     private void collectPlatforms(String quarkusCoreVersion, ExtensionCatalogBuilder catalogBuilder,
-            Collection<String> upstreamQuarkusVersions, RegistryExtensionResolver registry, int platformIndex,
+            RegistryExtensionResolver registry, int platformIndex,
             Platform p) throws RegistryResolutionException {
         for (PlatformStream s : p.getStreams()) {
             int releaseIndex = 0;
@@ -741,10 +734,9 @@ public class ExtensionCatalogResolver {
                 }
 
                 final String upstreamQuarkusVersion = r.getUpstreamQuarkusCoreVersion();
-                if (upstreamQuarkusVersion != null) {
-                    if (!upstreamQuarkusVersions.contains(upstreamQuarkusVersion)) {
-                        upstreamQuarkusVersions.add(upstreamQuarkusVersion);
-                    }
+                if (upstreamQuarkusVersion != null
+                        && !catalogBuilder.upstreamQuarkusVersions.contains(upstreamQuarkusVersion)) {
+                    catalogBuilder.upstreamQuarkusVersions.add(upstreamQuarkusVersion);
                 }
             }
         }

